src/hotspot/share/gc/shared/arraySlicer.hpp

```diff
@@ -0,0 +1,270 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_GC_SHARE_ARRAYSLICER_HPP
+#define SHARE_GC_SHARE_ARRAYSLICER_HPP
+
+#include "oops/oopsHierarchy.hpp"
+
+// A task queue entry that encodes both regular oops, and the array oops plus sliceing data for
+// parallel array processing.
+// The design goal is to make the regular oop ops very fast, because that would be the prevailing
+// case. On the other hand, it should not block parallel array processing from efficiently dividing
+// the array work.
+//
+// The idea is to steal the bits from the 64-bit oop to encode array data, if needed. For the
+// proper divide-and-conquer strategies, we want to encode the "blocking" data. It turns out, the
+// most efficient way to do this is to encode the array block as (slice * 2^pow), where it is assumed
+// that the block has the size of 2^pow. This requires for pow to have only 5 bits (2^32) to encode
+// all possible arrays.
+//
+//    |xx-------oop---------|-pow-|--slice---|
+//    0                    49     54        64
+//
+// By definition, slice == 0 means "no slice", i.e. sliceing starts from 1.
+//
+// This encoding gives a few interesting benefits:
+//
+// a) Encoding/decoding regular oops is very simple, because the upper bits are zero in that task:
+//
+//    |---------oop---------|00000|0000000000| // no slice data
+//
+//    This helps the most ubiquitous path. The initialization amounts to putting the oop into the word
+//    with zero padding. Testing for "slicedness" is testing for zero with slice mask.
+//
+// b) Splitting tasks for divide-and-conquer is possible. Suppose we have slice <C, P> that covers
+// interval [ (C-1)*2^P; C*2^P ). We can then split it into two slices:
+//      <2*C - 1, P-1>, that covers interval [ (2*C - 2)*2^(P-1); (2*C - 1)*2^(P-1) )
+//      <2*C, P-1>,     that covers interval [ (2*C - 1)*2^(P-1);       2*C*2^(P-1) )
+//
+//    Observe that the union of these two intervals is:
+//      [ (2*C - 2)*2^(P-1); 2*C*2^(P-1) )
+//
+//    ...which is the original interval:
+//      [ (C-1)*2^P; C*2^P )
+//
+// c) The divide-and-conquer strategy could even start with slice <1, round-log2-len(arr)>, and split
+//    down in the parallel threads, which alleviates the upfront (serial) splitting costs.
+//
+// Encoding limitations caused by current bitscales mean:
+//    10 bits for slice: max 1024 blocks per array
+//     5 bits for power: max 2^32 array
+//    49 bits for   oop: max 512 TB of addressable space
```

Comment:
I think hardware support is an orthogonal issue. It would have been an issue if we just blindly casted the pointer to `oop`, and relied on hardware to only treat the lowest bits for the actual address.

But since we are masking out the `oop` explicitly (see `oop_extract_mask`), we do not actually care what hardware supports. What we get as the limit is how much of the Java heap we can represent, given that we encode `oop`-s. Shenandoah checks this on startup, for example: https://github.com/openjdk/jdk/blob/d8c1c6ab0543c986280dcfa1c6c79e010a7b35fb/src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp#L202-L212

Reply:
What is a practical heap size?  512T should be enough for everyone :)

As a former colleague liked to say, don't try to stuff two bits of information
into one bit of data.